### ISSUE TITLE
[corgan-pr-prep] T2: Create CorGANGenerationMIMIC3 task function

### ISSUE THREAD

**Issue by jalengg** [jalengg/PyHealth#26]
## Spec

Create `pyhealth/tasks/corgan_generation.py` with:

1. `CorGANGenerationMIMIC3(BaseTask)`
   - `task_name = "CorGANGenerationMIMIC3"`
   - `input_schema = {"visits": "nested_sequence"}`
   - `output_schema = {}` (generative, no conditioning label)
   - `_icd_col = "diagnoses_icd/icd9_code"` (MIMIC-III)
   - `__call__(self, patient)` extracts ICD codes via polars, returns `[{"patient_id": ..., "visits": [[codes...], ...]}]` or `[]` for < 2 visits

2. `CorGANGenerationMIMIC4(CorGANGenerationMIMIC3)`
   - Overrides ONLY `task_name` and `_icd_col = "diagnoses_icd/icd_code"`
   - Inherits ALL other logic (DRY)

3. Module-level instances:
   - `corgan_generation_mimic3_fn = CorGANGenerationMIMIC3()`
   - `corgan_generation_mimic4_fn = CorGANGenerationMIMIC4()`

4. All 4 names exported from `pyhealth/tasks/__init__.py`

## Acceptance criteria

- Spec compliance: ✅ (all 14 requirements met)
- Code quality: pending (import order, Returns indentation, __init__.py placement)
- Pattern mirrors halo_generation.py exactly

## Status

Implementation committed (5280d46d). Quality fixes pending.

**Comment by jalengg** [comment]
## Why this task exists

The task function is the PyHealth 2.0 abstraction layer that decouples *what data looks like in the raw dataset* from *what the model needs as input*. Without it, every model must contain its own bespoke MIMIC-III parsing logic — duplicated, untested, and impossible to swap out.

When the original CorGAN code was written, it used a custom `CorGANDataset` and `_build_global_vocab` that directly consumed numpy arrays from preprocessed CSV files. That approach works for one-off experiments but is incompatible with the PyHealth ecosystem: it cannot be applied to MIMIC-IV with a one-line change, it cannot be used with PyHealth's `split_by_patient`, and it cannot be evaluated against other models trained on the same task.

The `CorGANGenerationMIMIC3` task function solves all three problems at once.

## What a task function actually is

In PyHealth 2.0, a `BaseTask` subclass is a *patient-level mapping function*. It receives a single `patient` object from the raw dataset and returns a list of structured sample dicts. The framework then applies this function across all patients to produce a `SampleDataset` — a flat list of samples that any model can consume.

For generative models, the output schema is deliberately empty (`output_schema = {}`). There is no label to predict. The model's job is to learn the joint distribution of the input features. The task function's only job is to extract the right input features.

## Why the MIMIC-III / MIMIC-IV split matters

MIMIC-III stores ICD diagnosis codes in a column named `icd9_code`. MIMIC-IV renamed it to `icd_code` (removing the version prefix, since MIMIC-IV includes both ICD-9 and ICD-10 codes). This is a schema difference, not a logic difference.

The DRY pattern — `CorGANGenerationMIMIC4` inherits all logic from `CorGANGenerationMIMIC3` and overrides only `_icd_col` — means that any bug fix or improvement to the extraction logic is automatically inherited by both variants. This is not just style preference; it prevented a real divergence bug during the HALO integration where MIMIC-IV received an outdated extraction path after a fix was applied only to the MIMIC-III class.

## Why we require at least 2 visits

CorGAN is trained on patient records to learn co-occurrence patterns of diagnoses. A patient with only one visit provides minimal training signal and, more importantly, synthetic patients generated by a CorGAN trained on single-visit records would have no temporal structure at all — they would be indistinguishable from random code sampling. The `< 2` visit filter is a data quality gate, not an arbitrary restriction. It matches the filtering logic established in the HALO integration for the same reason.

## Why polars, not pandas

PyHealth 2.0 uses polars for all internal data operations because it is significantly faster than pandas on large datasets (MIMIC-III has ~40k patients, ~500k admissions). The `.select(pl.col(self._icd_col)).to_series().drop_nulls().to_list()` chain is the canonical pattern established across all PyHealth task functions. Using pandas here would create an inconsistency that future maintainers would need to reconcile.

## The convenience instance pattern

`corgan_generation_mimic3_fn = CorGANGenerationMIMIC3()` at module level lets users write:

```python
sample_dataset = base_dataset.set_task(corgan_generation_mimic3_fn)
```

instead of:

```python
sample_dataset = base_dataset.set_task(CorGANGenerationMIMIC3())
```

This is purely ergonomic, but it follows the convention established by every other PyHealth task function. Consistency here matters because users copy-paste from examples and documentation.

**Comment by jalengg** [comment]
## Schema change: nested_sequence → multi_hot

During T3 smoke testing, we investigated whether `nested_sequence` was the right processor for CorGAN and concluded it was not. This comment documents the reasoning and the resulting change to T2.

### Why nested_sequence was wrong for CorGAN

`NestedSequenceProcessor` is designed for temporal models (transformers, RNNs) that consume visit sequences as ordered inputs. It outputs a 2D padded integer tensor of shape `(num_visits, max_codes_per_visit)`. CorGAN never uses this structure — it immediately discards visit boundaries and collapses all codes for a patient into a single flat binary vector.

Using `nested_sequence` created a semantic lie: the task function claimed to produce structured visit data, but the model silently threw the structure away on every training step. It also forced a manual `_encode_samples_to_multihot` method in the model that duplicated work already done by the processor.

### Why multi_hot is correct

`MultiHotProcessor` (schema key `"multi_hot"`) takes a flat list of code strings and produces a pre-built binary vector of shape `(vocab_size,)`, dtype float32. This is exactly what CorGAN trains on — no further encoding needed. It removes the need for `_encode_samples_to_multihot` entirely and lets `torch.stack` batch patients without a custom collate function.

### What changed in this task function

- `input_schema = {"visits": "multi_hot"}` (was `"nested_sequence"`)
- `__call__` now accumulates all ICD codes across all admissions into a **flat list** using `extend` instead of building a nested list with `append`
- The visit-count guard (`if len(visits) < 2`) was replaced with a code-presence guard (`if not all_codes`) — since visit structure no longer exists, filtering on zero codes is the correct invariant
- Returns `{"visits": ["ICD1", "ICD2", "ICD3"]}` (flat) instead of `{"visits": [["ICD1", "ICD2"], ["ICD3"]]}` (nested)

### Tradeoff acknowledged

With `multi_hot`, the task output is no longer interchangeable with HALO's `nested_sequence` output. This is intentional — CorGAN and HALO model fundamentally different data structures, and having the same schema would have been misleading. The schema difference makes the incompatibility explicit at the task level rather than hiding it inside the model.

**Comment by jalengg** [comment]
## T2 Complete ✅

**What was implemented:**
- New file: `pyhealth/tasks/corgan_generation.py` (98 lines)
- `CorGANGenerationMIMIC3(BaseTask)` with `input_schema = {"visits": "multi_hot"}`
- `CorGANGenerationMIMIC4` override for ICD-10 column name
- Module-level instances: `corgan_generation_mimic3_fn`, `corgan_generation_mimic4_fn`
- Registered in `pyhealth/tasks/__init__.py` (alphabetical order)

**Key design decision — `multi_hot` schema:**
During T3 implementation, a schema investigation found that `nested_sequence` (the initial choice) was semantically wrong for CorGAN. CorGAN is a bag-of-codes model that operates on flat binary patient vectors — it has no use for visit temporal structure. `MultiHotProcessor` (schema `"multi_hot"`) is the correct fit: it takes a flat list of code strings and produces a `(vocab_size,)` float32 binary tensor, exactly what CorGAN expects. This eliminated the need for a manual `_encode_samples_to_multihot` step entirely.

Task function logic:
- Flattens all ICD codes across all admissions into a single flat list per patient
- Returns `[]` if patient has no codes (early return)
- Returns `[{"patient_id": ..., "visits": [code1, code2, ...]}]` — one record per patient

**Commits:** `5280d46d` (initial), `544945d1` (quality fixes: import order, Returns indentation), `c02f1584` (schema switch to multi_hot), `77bfd406` (final quality fixes)

**Reviews:** Spec ✅ | Code quality ✅
